### ISSUE TITLE
Random Avatars Missing After Migration

### ISSUE THREAD

**Issue by dustinwilson** [tootsuite/mastodon#7967]
I run my own single user instance using the latest release, not on master. I originally ran Mastodon in Docker, and last Friday Docker decided to take a dirt nap on its own without any intervention from me and take Mastodon with it. Thankfully I had a backup of the database so I just migrated it to a vm as a regular installation. I didn't run into any issues there. Now, that I decided to move it back to my server as a regular installation everything's gone fine except random avatars are missing that were present on my vm. When it downloads new ones they're saved just fine but with 660 permissions, preventing nginx from viewing it.

I can go through `system/accounts/avatars` on my vm's installation and find the avatar just fine but the name of the file will be different. For example say I have `system/accounts/avatars/000/001/513/original/5f877bab88af5016.png` that's returning a 404. On my vm it would be something like `system/accounts/avatars/000/001/513/original/533c8e444f898db8.png`. I'm not sure what happened in-between the migration to cause this, and it's not all the avatars doing this.

As for the permissions thing anytime a new avatar is downloaded it has 660 applied, nothing like the rest of the system folder. Downloaded image attachments are just fine.

I have run `rake mastodon:media:redownload_avatars`, and while it thinks for a while it doesn't cause anything to be redownloaded that was missing anyway.

**Comment by dustinwilson** [comment]
I've come up with a tempfix for this. `redownload_avatars` doesn't fix this because the images won't show up in its list of ones to cull, so you have to manually redownload:

```
cd ~live
RAILS_ENV=production bundle exec rails c
a = Account.find_remote('USERNAME', 'REMOTE.DOMAIN')
a.reset_avatar!
a.reset_header!
a.save
```

Still doesn't explain why the avatars messed up in the first place.

**Comment by hugogameiro** [comment]
OK, so in my situation I don't think it was the same problem. I found that the users that `rake mastodon:media:redownload_avatars` skipped were the ones that had the value `avatar_remote_url`  blank in the database (also skipped header if `header_remote_url` was blank).

**Comment by skid9000** [comment]
I have the same problem on my instance ...

The DOMAIN argument dosen't work with the rake task, it create no jobs in sidekiq ...
But if i run the task without the domain argument, it works, no errors at all, but it only download like ... 2/3 of all avatars ...

So i had to refresh manualy some avatars from the admin panel ...
And some of them, even aftr refreshing, are still missing, so i had to download them manually ...

**Comment by Tagadda** [comment]
There is a topic about that (I think ?) on discourse right now : https://discourse.joinmastodon.org/t/fixing-missing-remote-avatars/801

**Comment by dustinwilson** [comment]
@hugogameiro `rake mastodon:media:redownload_avatars` did literally nothing in my situation. Sidekiq tasks weren't created so there were nothing in the logs. The hack I run is exactly the same thing `mastodon:media:redownload_avatars` does except just on one item at a time. They were being skipped in my situation somehow as well, but `avatar_remote_url` and `header_remote_url` were present. To solve the issue I pretty much just created a hacked-together cronjob that every 30 minutes grabs the past 30 minutes of 404 errors in the logs and redownloads those avatars and images (I was having a similar issue with image attachments, too) then logs the images it redownloads. The longer time passes the fewer avatars that were messed up in the migration are popping up, so eventually I'll turn it off.

@Tagadda Yeah, it's a similar issue and what helped me discover a temporary fix to mine.

**Comment by serenitylabs-sascha** [comment]
We've experienced this issue. I believe it might be related to low ulimit (1024 max files open). Some failed jobs in Sidekiq seem to indicate that. Something might be leaking file descriptors.

**Comment by damongant** [comment]
I believe this is still an actual error condition that needs to be detected.

**Comment by Gargron** [comment]
Yes, sidekiq creating a lot of jobs had some issues, but the task is now a tootctl command which has a background/foreground option, so it should work much better.
